[DAL] Extend algorithm anatomy (#212)
oneapi-src/oneAPI-spec: source/elements/oneDAL/source/programming_model/operations/index.rst

.. _operations:

==========
Operations
==========

**An operation** is a function that transforms :txtref:`a descriptor
<descriptors>` and other arguments represented via `an input <input_>`_ object

> to `a result <result_>`_ object. An operation responsible for:

Reviewer: ```suggestion
to `a result <result_>`_ object. An operation is responsible for:
```

> - Execution of all computational routines of the algorithm represented by the

Reviewer: probably better  + to align with other items in the list

```suggestion
- Executing all of an algorithm's computational routines represented by the
```

  descriptor.

- Passing SYCL* queue to computational routines.

> - Verification of preconditions and postconditions, before and after execution

Reviewer: ```suggestion
- Verifying preconditions and postconditions before and after the execution
```

  of computational routines.

.. _general_operation:

----------------------------
General operation definition
----------------------------
`The following code sample <operation_template_>`_ shows the declaration for an
abstract operation. To declare a particular operation, the ``%OPERATION_NAME%``
shall be substituted.

.. code-block:: cpp
   :name: operation_template

   namespace oneapi::dal {

   template <typename Descriptor>
   using %OPERATION_NAME%_input_t = /* implementation defined */;

   template <typename Descriptor>
   using %OPERATION_NAME%_result_t = /* implementation defined */;

   template <typename Descriptor>
   %OPERATION_NAME%_result_t<Descriptor> %OPERATION_NAME%(
      sycl::queue& queue,
      const Descriptor& desc,
      const %OPERATION_NAME%_input_t<Descriptor>& input);

   } // namespace oneapi::dal

Each operation shall satisfy the following requirements:

> - An operation shall accept tree parameters in the following order:

Reviewer 2: three?

  + The SYCL* queue object.
  + The descriptor of the algorithm.
  + The `input object <input_>`_.

- An operation shall return the `result object <result_>`_.

- The ``%OPERATION_NAME%_input_t`` and ``%OPERATION_NAME%_result_t`` alias
  templates shall be used for inference of the input and return types.

> - If precondition is violated, an operation shall throw an exception derived
>   from ``oneapi::dal::logic_error``.
> 
> - If postcondition is violated, an operation shall throw an exception derived
>   from ``oneapi::dal::runtime_error``.

Reviewer: ```suggestion
- If a precondition is violated, an operation shall throw an exception derived
  from ``oneapi::dal::logic_error``.

- If a postcondition is violated, an operation shall throw an exception derived
  from ``oneapi::dal::runtime_error``.
```

- The exact list of pre- and post- conditions shall be defined by :txtref:`a
  particular algorithm specification <algorithms>`.

-------------------
Operation shortcuts
-------------------
In order to make the code on user side less verbose, oneDAL defines the
following overloaded functions called *shortcuts* for each operation in addition
to the general one described in section `General operation definition
<general_operation_>`_.

> - Shortcut for execution on :term:`host <Host/Device>`. Performs the same
>   operation as the general function on host, but does not require passing the
>   queue explicitly.

Reviewer: ```suggestion
- A shortcut for execution on :term:`host <Host/Device>` that performs the same
  operation as the general function on host, but does not require the
  queue to be passed explicitly.
```

  .. code-block:: cpp

     template <typename Descriptor>
     %OPERATION_NAME%_result_t<Descriptor> %OPERATION_NAME%(
        const Descriptor& desc,
        const %OPERATION_NAME%_input_t<Descriptor>& input);

> - Shortcut that allows omitting explicit input creation.

Reviewer: ```suggestion
- A shortcut that allows omitting explicit input creation.
```

  .. code-block:: cpp

     template <typename Descriptor, typename... Args>
     %OPERATION_NAME%_result_t<Descriptor> %OPERATION_NAME%(
        sycl::queue& queue,
        const Descriptor& desc,
        Args&&... args);

> - Shortcut that allows omitting explicit queue and input creation. This is
>   combination of two previous shortcuts.

Reviewer: ```suggestion
- A shortcut that allows omitting explicit queue and input creation. This is a
  combination of two previous shortcuts.
```

.. code-block:: cpp
   :name: operation_template

   namespace oneapi::dal {

   template <typename Descriptor>
   using %OPERATION%_input_t = /* implementation defined */;

   template <typename Descriptor>
   using %OPERATION%_result_t = /* implementation defined */;

   template <typename Descriptor>
   %OPERATION%_result_t<Descriptor> %OPERATION%(
      sycl::queue& queue,
      const Descriptor& desc,
      const %OPERATION%_input_t<Descriptor>& input);

   } // namespace oneapi::dal

Each operation shall satisfy the following requirements:

- An operation shall accept three parameters in the following order:

  + The SYCL* queue object
  + The descriptor of the algorithm
  + The `input object <input_>`_

- An operation shall return the `result object <result_>`_.

- The ``%OPERATION%_input_t`` and ``%OPERATION%_result_t`` alias
  templates shall be used for inference of the input and return types.

- If a precondition is violated, an operation shall throw an exception derived
  from ``oneapi::dal::logic_error``.

- If a postcondition is violated, an operation shall throw an exception derived
  from ``oneapi::dal::runtime_error``.

> - If descriptor is incompatible with some operation, the error shall be reported

Reviewer: The descriptor, an error

following overloaded functions called *shortcuts* for each operation in addition
to the general one described in section `General operation definition
<general_operation_>`_.

- A shortcut for execution on :term:`host <Host/Device>` that performs the same
  operation as the general function on host, but does not require the
  queue to be passed explicitly.

  .. code-block:: cpp

     template <typename Descriptor>
     %OPERATION%_result_t<Descriptor> %OPERATION%(
        const Descriptor& desc,
        const %OPERATION%_input_t<Descriptor>& input);

- A shortcut that allows omitting explicit input creation.

  .. code-block:: cpp

     template <typename Descriptor, typename... Args>
     %OPERATION%_result_t<Descriptor> %OPERATION%(
        sycl::queue& queue,
        const Descriptor& desc,
        Args&&... args);

- A shortcut that allows omitting explicit queue and input creation. This is a
  combination of two previous shortcuts.

  .. code-block:: cpp

     template <typename Descriptor, typename... Args>
     %OPERATION%_result_t<Descriptor> %OPERATION%(
        const Descriptor& desc,
        Args&&... args);

.. _input:

-----
Input
-----

> An input objects aggregates all the data that algorithm requires for performing

Reviewer: object, the algorithm

> specific operation. The data represented via :txtref:`tables <table>`, so,

Reviewer: a specific operation, is represented

> typically, an input is a collection of tables, but not limited with them and can

Reviewer: limited to them

I'd split this sentence and say: An input can also aggregate...

> aggregate objects of arbitrary type.

Reviewer: An arbitrary type

Result
------

A result object aggregates all output values computed by the algorithm. All

> assumptions about `an input <input_>`_ is applied to a result.

Reviewer: ```suggestion
assumptions about `an input <input_>`_ are applied to a result as well.
```

   } // namespace oneapi::dal::%ALGORITHM%

-------------------
Supported operation
-------------------

> Please refer to the :txtref:`Supported operations <op_supported>` section for

Reviewer: ```suggestion
Refer to the :txtref:`Supported operations <op_supported>` section for
```
